File: frontend/src/lib/passwordPolicy.js
/**
 * Pravidla hesla — musí odpovídat backendu `password_policy.password_strength_error`.
 * Min. 8 znaků, písmeno, číslice, speciální znak (ASCII interpunkce).
 */

const SPECIAL_RE = /[!"#$%&'()*+,\-./:;<=>?@[\\\]^_`{|}~]/;

/**
 * @param {string} password
 * @returns {{ ok: boolean, message: string }}
 */
export function validatePasswordClient(password) {
  const p = password == null ? "" : String(password);
  if (p.length < 8) {
    return { ok: false, message: "Heslo musí mít alespoň 8 znaků." };
  }
  if (!/[A-Za-z]/.test(p)) {
    return { ok: false, message: "Heslo musí obsahovat alespoň jedno písmeno." };
  }
  if (!/\d/.test(p)) {
    return { ok: false, message: "Heslo musí obsahovat alespoň jednu číslici." };
  }
  if (!SPECIAL_RE.test(p)) {
    return { ok: false, message: "Heslo musí obsahovat alespoň jeden speciální znak (např. !@#$%)." };
  }
  return { ok: true, message: "" };
}

/** Krátký text do UI (registrace / nastavení). */
export const PASSWORD_POLICY_HINT =
  "Heslo: min. 8 znaků, alespoň jedno písmeno, jedna číslice a jeden speciální znak (!\"#$%…).";

File: backend-java/src/main/java/cz/bankintel/security/PasswordPolicy.java
package cz.bankintel.security;

import java.util.regex.Pattern;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ResponseStatusException;

/** Společná pravidla síly hesla — registrace, admin vytvoření uživatele, změna hesla. */
@Component
public class PasswordPolicy {

    private static final Pattern LETTER = Pattern.compile("[A-Za-z]");
    private static final Pattern DIGIT = Pattern.compile("\\d");
    private static final Pattern SPECIAL = Pattern.compile("[!\"#$%&'()*+,\\-./:;<=>?@\\[\\\\\\]^_`{|}~]");

    public void validateOrThrow(String password) {
        String message = strengthError(password);
        if (message != null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, message);
        }
    }

    static String strengthError(String password) {
        if (password == null) {
            return "Heslo je povinné.";
        }
        if (password.length() < 8) {
            return "Heslo musí mít alespoň 8 znaků.";
        }
        if (!LETTER.matcher(password).find()) {
            return "Heslo musí obsahovat alespoň jedno písmeno.";
        }
        if (!DIGIT.matcher(password).find()) {
            return "Heslo musí obsahovat alespoň jednu číslici.";
        }
        if (!SPECIAL.matcher(password).find()) {
            return "Heslo musí obsahovat alespoň jeden speciální znak (např. !@#$%).";
        }
        return null;
    }
}
